src/MongoDB.Driver.Core/Core/Compression/CompressorFactory.cs

```diff
@@ -0,0 +1,117 @@
+{
+    /// <summary>
+    /// Represents a compressor factory.
+    /// </summary>
+    public interface ICompressorFactory
+    {
+        /// <summary>
+        /// Gets or creates a compressor based on the compressor type.
+        /// </summary>
+        /// <param name="compressorType">The compressor type.</param>
+        /// <returns>The compressor.</returns>
+        ICompressor GetOrCreate(CompressorType compressorType);
+    }
+
+    internal class CompressorFactory : ICompressorFactory
+    {
+        private readonly Dictionary<CompressorType, ICompressor> _cache;
+        private readonly IReadOnlyList<ConnectionStringCompressorDescription> _allowedCompressors;
+
+        public CompressorFactory(IReadOnlyList<ConnectionStringCompressorDescription> allowedCompressors)
+        {
+            _cache = new Dictionary<CompressorType, ICompressor>();
+            _allowedCompressors = Ensure.IsNotNull(allowedCompressors, nameof(allowedCompressors));
+        }
+
+        public ICompressor GetOrCreate(CompressorType compressorType)
+        {
+            if (_cache.TryGetValue(compressorType, out var cachedCompressor))
+            {
+                return cachedCompressor;
+            }
+            else
+            {
+                var requestedCompressor = _allowedCompressors.FirstOrDefault(c => c.Type == compressorType);
+                if (requestedCompressor != null)
+                {
+                    var newCompressor = CreateCompressor(requestedCompressor);
+                    if (newCompressor != null)
+                    {
+                        _cache.Add(compressorType, newCompressor);
+                        return newCompressor;
+                    }
+                }
+
+                if (compressorType == CompressorType.Noop)
+                {
+                    return null;
+                }
+
+                throw new ArgumentException($"The compressor type '{compressorType}' is not allowed.", nameof(compressorType));
+            }
+        }
+
+        private ICompressor CreateCompressor(ConnectionStringCompressorDescription compressorDescription)
+        {
+            switch (compressorDescription.Type)
+            {
+                case CompressorType.Zlib:
+                    {
+                        int? zlibCompressionLevel = null;
+                        if (compressorDescription.Properties.ContainsKey("Level"))
+                        {
+                            zlibCompressionLevel = (int)compressorDescription.Properties["Level"];
+                        }
+
+                        return new ZlibCompressor(zlibCompressionLevel);
+                    }
+                case CompressorType.Snappy:
+                    return new SnappyCompressor();
+            }
+
+            return null;
+        }
+    }
+
+    internal static class CompressorTypeExtensions
+    {
+        public static bool IsAvailable(this CompressorType compressorType)
```

Comment:
The idea here was to not push `CompressorFactory` to low-level steps(like parsing connection strings and so on) at least explicitly since on that level we don't need to create/or use compressors. 
Also, in case if we use a new possible helper class for that target, I guess it will make code harder to read since we need to know which helper we need to use for that target, in my case we need just to call intellisense. 

So, why do you think the extension method is not suitable here?